### PR TITLE
Add a portable runner that renders pipelines as a dot graph.

### PR DESCRIPTION
**Pull request by robertwb**
This runner can either render the graph to a (set of) output path(s), as
designated by (possibly repeated) --render_output, or serve the pipeline as
an interactive graph, if --render_port is set.

In Python, this runner can be passed directly at pipeline construction, e.g.::

   with beam.Pipeline(runner=beam.runners.render.RenderRunner(), options=...)

For other languages, start this service a by running::

  python -m apache_beam.runners.render --job_port=PORT ...

and then run your pipline with the PortableRunner setting the job endpoint
to `localhost:PORT`.

If any `--render_output=path.ext` flags are passed, each submitted job will
get written to the given output (overwriting any previously existing file).

If `--render_port` is set to a non-negative value, a local http server will
be started which allows for interactive exploration of the pipeline graph.

Requires the that graphviz dot executable be available in the path.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by robertwb** [comment]
R: @tvalentyn

**Comment by tvalentyn** [comment]
> input would be a pipeline.pb file

or .textproto

**Comment by robertwb** [comment]
This works entirely on the proto, and can be run as a portable runner. I was thinking about accepting a proto file as input as well (though it's not clear how to get the proto out for many SDKs).

File: sdks/python/apache_beam/pipeline.py [hunk 1]
```diff
@@ -1285,7 +1285,8 @@ def named_inputs(self):
           self.main_inputs, self.side_inputs)
       if not self.parts:
         for name, pc_out in self.outputs.items():
-          if pc_out.producer is not self:
+          if pc_out.producer is not self and pc_out not in named_inputs.values(
```
Comment (kennknowles) [inline sdks/python/apache_beam/pipeline.py:1288]: Is this a bug fix that might be best served as a separate commit? Etc elsewhere

File: sdks/python/apache_beam/runners/render.py [hunk 2]
```diff
@@ -0,0 +1,529 @@
...
+                       input=layout,
+                       capture_output=True,
+                       check=True).stdout)
+    for box in json_out['objects']:
+      name = box.get('name', None)
+      if box.get('name', None) in self.pipeline.components.transforms:
+        if 'pos' in box:
+          self.latest_positions[name] = box['pos']
+        elif 'bb' in box:
+          x0, y0, x1, y1 = [float(r) for r in box['bb'].split(',')]
```
Comment (tvalentyn) [inline sdks/python/apache_beam/runners/render.py:305]: lint fails here due to unused variables.

**Review by tvalentyn — COMMENTED** [review]
I added some additional functionality to consider, feel free to drop/edit as necessary. didn't take a close look at the code yet, but works great.

File: sdks/python/apache_beam/pipeline.py (hunk 1 above)
Comment (robertwb) [reply sdks/python/apache_beam/pipeline.py:1288]: This resulted in some spurious edges that didn't impact anything but rendering. (The other change was just a refactoring to re-use the side inputs code.)

File: sdks/python/apache_beam/runners/render.py (hunk 2 above)
Comment (robertwb) [reply sdks/python/apache_beam/runners/render.py:309]: Fixed.

**Review by robertwb — COMMENTED** [review]
Thanks. Your changes look fine.

File: sdks/python/apache_beam/runners/render.py [hunk 3]
```diff
@@ -0,0 +1,535 @@
...
+        not self.pipeline.components.transforms[transform_id].subtransforms)
+
+  def info(self):
+    if len(self.highlighted) != 1:
+      return ''
+    transform_id = self.highlighted[0]
+    return f'<pre>{self.pipeline.components.transforms[transform_id]}</pre>'
+
+  def layout_dot(self):
+    with open('out.dot', 'w') as fout:
```
Comment (tvalentyn) [inline sdks/python/apache_beam/runners/render.py:290]: is this a leftover? if not, I am not seeing where this is used and should probably be a tempfile.

File: sdks/python/apache_beam/runners/render.py [hunk 4]
```diff
@@ -0,0 +1,535 @@
...
+
+    # Try to capture the positions for layout consistency.
+    json_out = json.loads(
+        subprocess.run(['dot', '-n2', '-Kneato', '-Tjson'],
+                       input=layout,
+                       capture_output=True,
+                       check=True).stdout)
+    for box in json_out['objects']:
+      name = box.get('name', None)
+      if box.get('name', None) in self.pipeline.components.transforms:
```
Comment (tvalentyn) [inline sdks/python/apache_beam/runners/render.py:305]:
```suggestion
      if name in self.pipeline.components.transforms:
```

File: sdks/python/apache_beam/runners/render.py [hunk 5]
```diff
@@ -0,0 +1,535 @@
...
+If any `--render_output=path.ext` flags are passed, each submitted job will
+get written to the given output (overwriting any previously existing file).
+
+If `--render_port` is set to a non-negative value, a local http server will
+be started which allows for interactive exploration of the pipeline graph.
+
+As an alternative to starting a job server, a single pipeline can be rendered
+by passing a pipeline proto file to `--pipeline_proto`.
+
+Requires the that graphviz dot executable be available in the path.
```
Comment (tvalentyn) [inline sdks/python/apache_beam/runners/render.py:44]:
```suggestion
Requires the graphviz dot executable to be available in the path.
```

File: sdks/python/apache_beam/runners/render.py (hunk 2 above)
Comment (tvalentyn) [reply sdks/python/apache_beam/runners/render.py:309]: the expression is incorrectly quoted. did you mean to supply an f-string?

**Review by tvalentyn — APPROVED** [review]
Overall, LGTM thanks a lot for contributing this runner!

Left a few nits + 1 comment still to be addressed. Please doublecheck that rendering matches your expectations once `pos` attribute logic is fixed.
